task-maker-async/src/file_set.rs

```diff
@@ -0,0 +1,265 @@
+    Pending,
+    // The file set was created, but writing has not started yet.
+    Created,
+    // Writing has started but not yet finished.
+    Writing,
+    // Writing is complete.
+    Finalized,
+}
+
+#[derive(Debug)]
+pub struct FileSet {
+    kind: FileSetKind,
+    status: FileSetStatus,
+    writing_waiters: Vec<Sender<()>>,
+    creation_waiters: Vec<Sender<()>>,
+    finalization_waiters: Vec<Sender<()>>,
+    files: HashMap<FileSetFile, FileInfo>,
+}
+
+fn maybe_wait(
+    cond: bool,
+    waiters: &mut Vec<Sender<()>>,
+) -> impl Future<Output = Result<(), RecvError>> {
+    let receiver = if cond {
+        let (sender, receiver) = channel();
+        waiters.push(sender);
+        Some(receiver)
+    } else {
+        None
+    };
+    async {
+        if let Some(recv) = receiver {
+            recv.await
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl FileSet {
+    pub fn new() -> FileSet {
+        FileSet {
+            kind: FileSetKind::InputFile,
+            status: FileSetStatus::Pending,
+            writing_waiters: vec![],
+            creation_waiters: vec![],
+            finalization_waiters: vec![],
+            files: HashMap::new(),
+        }
+    }
+
+    pub fn create(&mut self, kind: FileSetKind) -> Result<bool, ()> {
+        if self.status > FileSetStatus::Pending {
+            if self.kind != kind {
+                return Err(());
+            }
+            return Ok(false);
+        }
+        self.kind = kind;
+        self.status = FileSetStatus::Created;
+        self.creation_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
+        Ok(true)
+    }
+
+    pub fn start_writing(&mut self) -> bool {
+        if self.status > FileSetStatus::Created {
+            return false;
+        }
+        self.status = FileSetStatus::Writing;
+        self.writing_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
+        true
+    }
+
+    pub fn mark_finalized(&mut self) -> bool {
+        if self.status > FileSetStatus::Writing {
+            return false;
+        }
+        self.status = FileSetStatus::Finalized;
+        self.finalization_waiters.drain(..).for_each(|waiter| {
+            let _ = waiter.send(());
+        });
+        for file in self.files.iter_mut() {
+            file.1.readers.drain(..).for_each(|mut waiter| {
+                let _ = waiter.send(FileReadingOutcome::from_data(
+                    &file.1.data,
+                    waiter.offset,
+                    waiter.chunk_size,
+                ));
+            });
+        }
+        true
+    }
+
+    pub fn append_to_file(&mut self, file: &FileSetFile, data: &[u8]) {
+        let file_data = self.files.entry(file.clone()).or_default();
+        file_data.data.extend_from_slice(data);
+        file_data.readers.drain(..).for_each(|mut waiter| {
```

Comment:
Can this function be called multiple times for the same file? If so, the readers here are consumed during the first call and they are not notified for later writes. Is this right?

Reply:
Yep, that's right: the purpose of this vector of readers is just to wake up readers that were waiting for a change on the file -- subsequent reads will be done directly, or re-insert new readers in the vector.